# Fifa practica

Volem predir el "value" dels jugadors utilitzant la informació del jugador.
## Importar llibreries

Importam les llibreries necesaries

In [1]:
import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

## Llegim les dades

Per llegir-les utilitzam ``pandas`` i la utilitat de la llibreria
``os``.


In [2]:
df = pd.read_csv(os.path.join("..", "in", "fifa.csv"))

## Anàlisi de dades

Aquestes són les nostres dades i hem de fer un anàlisi 

In [3]:
pd.set_option('display.max_columns', None)

In [4]:
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

# Eliminació de columnes
A aquesta part eliminaré les columnes que no aporten informació a la resolució del problema :
ID,Nationality,Photo,Preferred Foot,Name, Flag , Club Logo , Real Face , Joined , Jersey Number , Loaned From , Contract Valid Until , Weight , Height.

Totes aquestes columnes excepte "Joined" i "Loaned From" no són necesaries i es trivial.
Per altra banda, tampoc ens interesa conèixer l'any en que va entrar la persona al club perqué és independent del seu "Value".
Si és un jugador cedit tampoc ens interesa, perque hi ha molt pocs que ho són i el "Value" del jugador no dependrà d'ell.


Weight i Height són atributs que es podria realitzar un tractament. En el cas de Height seria canviar el x'x per x.x perque el programa ho pugui computar i en el cas de Weight seria llevar la part de Xlbs i pasar-ho a kg si es prefereix però he decidit no tractar-les perque són quasi irrelevants per calcular el "Value" del jugador ja que hi ha altres atributs que són més importants.

Work Rate i Body Type es podria fer un tractament similar al que realitzaré per "Club" que consisteix en agafar totes les files diferents i transformar-les en columnes i si el jugador té aquest atribut aleshores asignar-li un 1 i un 0 si no el té. Per Work Rate primer s'hauria de fer un tractament previ perque té 2 valors , un per la primera part i un altre per la segona, aleshores podriem agafar el de la primera part i l'altre eliminar-lo i a continuació realitzar el tractament anterior.Però per calcular el "Value" hi ha atributs més importants i aquests els podem eliminar.

Preferred foot: atribut amb molt poc impacte sobre "Value" que podem eliminar també.

In [5]:
df.drop(['ID','Preferred Foot','Name','Nationality','Body Type','Weight','Height','Work Rate','Photo','Flag','Club Logo','Real Face','Joined','Jersey Number','Loaned From','Contract Valid Until'],axis=1,inplace=True)
df

Unnamed: 0  Age  Overall  Potential                 Club    Value  \
0               0   31       94         94         FC Barcelona  €110.5M   
1               1   33       94         94             Juventus     €77M   
2               2   26       92         93  Paris Saint-Germain  €118.5M   
3               3   27       91         93    Manchester United     €72M   
4               4   27       91         92      Manchester City    €102M   
...           ...  ...      ...        ...                  ...      ...   
18202       18202   19       47         65      Crewe Alexandra     €60K   
18203       18203   19       47         63       Trelleborgs FF     €60K   
18204       18204   16       47         67     Cambridge United     €60K   
18205       18205   17       47         66      Tranmere Rovers     €60K   
18206       18206   16       46         66      Tranmere Rovers     €60K   

        Wage  Special  International Reputation  Weak Foot  Skill Moves  \
0      €565K     2202                       5.0        4.0          4.0   
1      €405K     2228                       5.0        4.0          5.0   
2      €290K     2143                       5.0        5.0          5.0   
3      €260K     1471                       4.0        3.0          1.0   
4      €355K     2281                       4.0        5.0          4.0   
...      ...      ...                       ...        ...          ...   
18202    €1K     1307                       1.0        2.0          2.0   
18203    €1K     1098                       1.0        2.0          2.0   
18204    €1K     1189                       1.0        3.0          2.0   
18205    €1K     1228                       1.0        3.0          2.0   
18206    €1K     1321                       1.0        3.0          2.0   

      Position    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0           RF  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1           ST  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2           LW  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3           GK   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4          RCM  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202       CM  42+2  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2   
18203       ST  45+2  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2   
18204       ST  45+2  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2   
18205       RW  47+2  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2   
18206       CM  43+2  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2   

        RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0      93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1      88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2      89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  45+2  44+2  45+2  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
18203  40+2  38+2  35+2  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2   
18204  44+2  44+2  38+2  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2   
18205  45+2  46+2  39+2  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2   
18206  45+2  46+2  45+2  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2   

        LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  \
0      47+2  47+2  47+2  59+2      84.0       95.0             70.0   
1      53+3  53+3  53+3  61+3      84.0       94.0             89.0   
2      47+3  47+3  47+3  60+3      79.0       87.0             62.0   

# Arreglar NaNs

Hi ha valors que són NaN,és a dir "not a number" ,valors nulls. Aquí hi ha les columnes que tenen algun NaN

In [6]:
df.columns[df.isna().any()].tolist()

['Club',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Position',
 'LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

Amb la següent instrucció podem veure la informació del nostre dataframe, les columnes Non-Null indiquen el total de jugadors que tenen un valor en aquell atribut. 


El total de jugadors són 18207 , per tant , haurem de tractar totes les columnes amb NaN.

Per altra banda, totes les columnes amb Dtype = object hauran de ser convertides a int per tal de poder ser interpretades.


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18207 entries, 0 to 18206
Data columns (total 73 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                18207 non-null  int64  
 1   Age                       18207 non-null  int64  
 2   Overall                   18207 non-null  int64  
 3   Potential                 18207 non-null  int64  
 4   Club                      17966 non-null  object 
 5   Value                     18207 non-null  object 
 6   Wage                      18207 non-null  object 
 7   Special                   18207 non-null  int64  
 8   International Reputation  18159 non-null  float64
 9   Weak Foot                 18159 non-null  float64
 10  Skill Moves               18159 non-null  float64
 11  Position                  18147 non-null  object 
 12  LS                        16122 non-null  object 
 13  ST                        16122 non-null  object 
 14  RS    

Ens podem fixar que hi ha columnes que tenen el mateix valor però que no és el total, aquest valor és 18159. És a dir,hi ha 48 jugadors dels quals no se sap una serie d'atributs, aquests es poden eliminar ja que 48 jugadors sobre 18207 no és un canvi gran. 

Procedim a eliminar-los (podem executar una instrucció que elimini els jugadors que no tenen Marking per exemple ja que seran els mateixos 48 que no tenen l'altra serie de atributs.

In [8]:
df = df[df['Marking'].notna()]
df

Unnamed: 0  Age  Overall  Potential                 Club    Value  \
0               0   31       94         94         FC Barcelona  €110.5M   
1               1   33       94         94             Juventus     €77M   
2               2   26       92         93  Paris Saint-Germain  €118.5M   
3               3   27       91         93    Manchester United     €72M   
4               4   27       91         92      Manchester City    €102M   
...           ...  ...      ...        ...                  ...      ...   
18202       18202   19       47         65      Crewe Alexandra     €60K   
18203       18203   19       47         63       Trelleborgs FF     €60K   
18204       18204   16       47         67     Cambridge United     €60K   
18205       18205   17       47         66      Tranmere Rovers     €60K   
18206       18206   16       46         66      Tranmere Rovers     €60K   

        Wage  Special  International Reputation  Weak Foot  Skill Moves  \
0      €565K     2202                       5.0        4.0          4.0   
1      €405K     2228                       5.0        4.0          5.0   
2      €290K     2143                       5.0        5.0          5.0   
3      €260K     1471                       4.0        3.0          1.0   
4      €355K     2281                       4.0        5.0          4.0   
...      ...      ...                       ...        ...          ...   
18202    €1K     1307                       1.0        2.0          2.0   
18203    €1K     1098                       1.0        2.0          2.0   
18204    €1K     1189                       1.0        3.0          2.0   
18205    €1K     1228                       1.0        3.0          2.0   
18206    €1K     1321                       1.0        3.0          2.0   

      Position    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0           RF  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1           ST  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2           LW  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3           GK   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4          RCM  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202       CM  42+2  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2   
18203       ST  45+2  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2   
18204       ST  45+2  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2   
18205       RW  47+2  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2   
18206       CM  43+2  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2   

        RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0      93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1      88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2      89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  45+2  44+2  45+2  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
18203  40+2  38+2  35+2  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2   
18204  44+2  44+2  38+2  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2   
18205  45+2  46+2  39+2  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2   
18206  45+2  46+2  45+2  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2   

        LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  \
0      47+2  47+2  47+2  59+2      84.0       95.0             70.0   
1      53+3  53+3  53+3  61+3      84.0       94.0             89.0   
2      47+3  47+3  47+3  60+3      79.0       87.0             62.0   

Eliminam els jugadors que no tenen Club , ja que són menys de 100 i no afecten als nostres càlculs.
Es podria fer de una altra manera canviant el NaN per *Sense Club* però amb la quantitat que hi ha he preferit eliminar-los.

In [9]:
df = df[df['Club'].notna()]
df

Unnamed: 0  Age  Overall  Potential                 Club    Value  \
0               0   31       94         94         FC Barcelona  €110.5M   
1               1   33       94         94             Juventus     €77M   
2               2   26       92         93  Paris Saint-Germain  €118.5M   
3               3   27       91         93    Manchester United     €72M   
4               4   27       91         92      Manchester City    €102M   
...           ...  ...      ...        ...                  ...      ...   
18202       18202   19       47         65      Crewe Alexandra     €60K   
18203       18203   19       47         63       Trelleborgs FF     €60K   
18204       18204   16       47         67     Cambridge United     €60K   
18205       18205   17       47         66      Tranmere Rovers     €60K   
18206       18206   16       46         66      Tranmere Rovers     €60K   

        Wage  Special  International Reputation  Weak Foot  Skill Moves  \
0      €565K     2202                       5.0        4.0          4.0   
1      €405K     2228                       5.0        4.0          5.0   
2      €290K     2143                       5.0        5.0          5.0   
3      €260K     1471                       4.0        3.0          1.0   
4      €355K     2281                       4.0        5.0          4.0   
...      ...      ...                       ...        ...          ...   
18202    €1K     1307                       1.0        2.0          2.0   
18203    €1K     1098                       1.0        2.0          2.0   
18204    €1K     1189                       1.0        3.0          2.0   
18205    €1K     1228                       1.0        3.0          2.0   
18206    €1K     1321                       1.0        3.0          2.0   

      Position    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0           RF  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1           ST  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2           LW  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3           GK   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4          RCM  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202       CM  42+2  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2   
18203       ST  45+2  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2   
18204       ST  45+2  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2   
18205       RW  47+2  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2   
18206       CM  43+2  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2   

        RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0      93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1      88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2      89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  45+2  44+2  45+2  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
18203  40+2  38+2  35+2  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2   
18204  44+2  44+2  38+2  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2   
18205  45+2  46+2  39+2  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2   
18206  45+2  46+2  45+2  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2   

        LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  \
0      47+2  47+2  47+2  59+2      84.0       95.0             70.0   
1      53+3  53+3  53+3  61+3      84.0       94.0             89.0   
2      47+3  47+3  47+3  60+3      79.0       87.0             62.0   

Hem de modificar els jugadors que tenen "Release Clause" = NaN per un valor igual a 0 ja que si el valor de la clàusula és NaN, significa que no en té i que és igual a 0 i el jugador es pot anar gratis.

In [10]:
df['Release Clause'] = df['Release Clause'].fillna('€0K')
df

Unnamed: 0  Age  Overall  Potential                 Club    Value  \
0               0   31       94         94         FC Barcelona  €110.5M   
1               1   33       94         94             Juventus     €77M   
2               2   26       92         93  Paris Saint-Germain  €118.5M   
3               3   27       91         93    Manchester United     €72M   
4               4   27       91         92      Manchester City    €102M   
...           ...  ...      ...        ...                  ...      ...   
18202       18202   19       47         65      Crewe Alexandra     €60K   
18203       18203   19       47         63       Trelleborgs FF     €60K   
18204       18204   16       47         67     Cambridge United     €60K   
18205       18205   17       47         66      Tranmere Rovers     €60K   
18206       18206   16       46         66      Tranmere Rovers     €60K   

        Wage  Special  International Reputation  Weak Foot  Skill Moves  \
0      €565K     2202                       5.0        4.0          4.0   
1      €405K     2228                       5.0        4.0          5.0   
2      €290K     2143                       5.0        5.0          5.0   
3      €260K     1471                       4.0        3.0          1.0   
4      €355K     2281                       4.0        5.0          4.0   
...      ...      ...                       ...        ...          ...   
18202    €1K     1307                       1.0        2.0          2.0   
18203    €1K     1098                       1.0        2.0          2.0   
18204    €1K     1189                       1.0        3.0          2.0   
18205    €1K     1228                       1.0        3.0          2.0   
18206    €1K     1321                       1.0        3.0          2.0   

      Position    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
0           RF  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2   
1           ST  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3   
2           LW  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3           GK   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4          RCM  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202       CM  42+2  42+2  42+2  44+2  44+2  44+2  44+2  44+2  45+2  45+2   
18203       ST  45+2  45+2  45+2  39+2  42+2  42+2  42+2  39+2  40+2  40+2   
18204       ST  45+2  45+2  45+2  45+2  46+2  46+2  46+2  45+2  44+2  44+2   
18205       RW  47+2  47+2  47+2  47+2  46+2  46+2  46+2  47+2  45+2  45+2   
18206       CM  43+2  43+2  43+2  45+2  44+2  44+2  44+2  45+2  45+2  45+2   

        RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
0      93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2   
1      88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3   
2      89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3   
3       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4      88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
18202  45+2  44+2  45+2  45+2  45+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2   
18203  40+2  38+2  35+2  35+2  35+2  38+2  30+2  31+2  31+2  31+2  30+2  29+2   
18204  44+2  44+2  38+2  38+2  38+2  44+2  34+2  30+2  30+2  30+2  34+2  33+2   
18205  45+2  46+2  39+2  39+2  39+2  46+2  36+2  32+2  32+2  32+2  36+2  35+2   
18206  45+2  46+2  45+2  45+2  45+2  46+2  46+2  46+2  46+2  46+2  46+2  46+2   

        LCB    CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  \
0      47+2  47+2  47+2  59+2      84.0       95.0             70.0   
1      53+3  53+3  53+3  61+3      84.0       94.0             89.0   
2      47+3  47+3  47+3  60+3      79.0       87.0             62.0   

Hi ha un problema amb els porters. Els porters no tenen puntuació a les columnes "LS","RT","ST",etc. i és lògic perquè un porter només juga de porter.

Per això totes les columnes que tenguin NaN al les columnes de "LS","RT","ST",etc. seran emplenades amb 0.

In [11]:
df['LS'] = df['LS'].fillna('0')
df['ST'] = df['ST'].fillna('0')
df['RS'] = df['RS'].fillna('0')
df['LW'] = df['LW'].fillna('0')
df['LF'] = df['LF'].fillna('0')
df['CF'] = df['CF'].fillna('0')
df['RF'] = df['RF'].fillna('0')
df['RW'] = df['RW'].fillna('0')
df['LAM'] = df['LAM'].fillna('0')
df['CAM'] = df['CAM'].fillna('0')
df['RAM'] = df['RAM'].fillna('0')
df['LM'] = df['LM'].fillna('0')
df['LCM'] = df['LCM'].fillna('0')
df['CM'] = df['CM'].fillna('0')
df['RCM'] = df['RCM'].fillna('0')
df['RM'] = df['RM'].fillna('0')
df['LWB'] = df['LWB'].fillna('0')
df['LDM'] = df['LDM'].fillna('0')
df['CDM'] = df['CDM'].fillna('0')
df['RDM'] = df['RDM'].fillna('0')
df['RWB'] = df['RWB'].fillna('0')
df['LB'] = df['LB'].fillna('0')
df['LCB'] = df['LCB'].fillna('0')
df['CB'] = df['CB'].fillna('0')
df['RCB'] = df['RCB'].fillna('0')
df['RB'] = df['RB'].fillna('0')

## Transformació de columnes a int
"Value" , "Wage" y "Release Clause" son columnes amb expressions que el programa no sap interpretar per això les hem de transformar

Utilitzarem el següent mètode:

In [12]:
def value_to_float(x):
    """
    From K and M to float.
    
    """
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

In [13]:
df["Value"] = df["Value"].apply(value_to_float)
df["Wage"] = df["Wage"].apply(value_to_float)
df["Release Clause"] = df["Release Clause"].apply(value_to_float)

Hem de modificar els valors de "LS	ST	RS	LW	LF	CF	RF	RW	LAM	CAM	RAM	LM	LCM	CM	RCM	RM	LWB	LDM	CDM	RDM	RWB	LB	LCB	CB	RCB	RB" perque duen un + i el programa no els sap interpretar.

Hem de crear el següent mètode que suma els dos valors. Exemple -> 88+2 = 90

In [14]:
#funció que suma el str per exemple: 88+2 = 90 i el transforma a int, sino és str significa que és un porter i és igual a 0
def skillSuma(valor):
    if type(valor) == str:
        s1 = valor[0:2]
        s2 = valor[-1] 
        valor = int(s1) +int(s2)
        return valor
    else:
        return 0

In [15]:

    skillColumnes = ['LS', 'ST', 'RS','LW','LF','CF','RF','RW','LAM','CAM',
                     'RAM','LM','LCM','CM','RCM','RM','LWB','LDM','CDM','RDM',
                     'RWB','LB','LCB','CB','RCB','RB']
    
    
    for columna in skillColumnes:
        df[columna] = df[columna].apply(skillSuma)

In [16]:
df.head()

Unnamed: 0  Age  Overall  Potential                 Club        Value  \
0           0   31       94         94         FC Barcelona  110500000.0   
1           1   33       94         94             Juventus   77000000.0   
2           2   26       92         93  Paris Saint-Germain  118500000.0   
3           3   27       91         93    Manchester United   72000000.0   
4           4   27       91         92      Manchester City  102000000.0   

       Wage  Special  International Reputation  Weak Foot  Skill Moves  \
0  565000.0     2202                       5.0        4.0          4.0   
1  405000.0     2228                       5.0        4.0          5.0   
2  290000.0     2143                       5.0        5.0          5.0   
3  260000.0     1471                       4.0        3.0          1.0   
4  355000.0     2281                       4.0        5.0          4.0   

  Position  LS  ST  RS  LW  LF  CF  RF  RW  LAM  CAM  RAM  LM  LCM  CM  RCM  \
0       RF  90  90  90  94  95  95  95  94   95   95   95  93   86  86   86   
1       ST  94  94  94  92  93  93  93  92   91   91   91  91   84  84   84   
2       LW  87  87  87  92  92  92  92  92   92   92   92  91   84  84   84   
3       GK   0   0   0   0   0   0   0   0    0    0    0   0    0   0    0   
4      RCM  85  85  85  90  90  90  90  90   91   91   91  91   90  90   90   

   RM  LWB  LDM  CDM  RDM  RWB  LB  LCB  CB  RCB  RB  Crossing  Finishing  \
0  93   66   63   63   63   66  61   49  49   49  61      84.0       95.0   
1  91   68   64   64   64   68  64   56  56   56  64      84.0       94.0   
2  91   68   63   63   63   68  63   50  50   50  63      79.0       87.0   
3   0    0    0    0    0    0   0    0   0    0   0      17.0       13.0   
4  91   80   80   80   80   80  76   69  69   69  76      93.0       82.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             21.0          50.0     13.0       18.0   21.0        19.0   
4             55.0          92.0     82.0       86.0   85.0        83.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         51.0         42.0          57.0         58.0     60.0       90.0   
4         91.0         91.0          78.0         76.0     79.0       91.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     43.0       31.0     67.0     43.0      64.0       12.0        38.0   
4     77.0       91.0     63.0     90.0      75.0       91.0        76.0   

   Interceptions  Positioning  Vision  Penalties  Composure  Marking  \
0           22.0         94.0    94.0       75.0       96.0     33.0   
1           29.0         95.0    82.0       85.0       95.0     28.0   
2           36.0         89.0    87.0       81.0       94.0     27.0   
3           30.0         12.0    68.0       40.0       68.0     15.0   
4           61.0         87.0    94.0       79.0       88.0     68.0   

   StandingTackle  SlidingTackle  GKDiving  GKHandling  GKKicking  \
0            28.0           26.0       6.0        11.0       15.0   
1            31.0           23.0       7.0        11.0       15.0   
2            24.0           33.0       9.0         9.0       15.0   
3            21.0           13.0      90.0        85.0       87.0   
4       

In [17]:
clb = df.pop("Club")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)

In [18]:
df['clb_1. FC Heidenheim 1846'].value_counts()

0    17890
1       28
Name: clb_1. FC Heidenheim 1846, dtype: int64

In [19]:
pos = df.pop("Position")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pos, prefix='pos').reset_index(drop=True)], axis=1, sort=False)

In [20]:
df.head()

Unnamed: 0  Age  Overall  Potential        Value      Wage  Special  \
0           0   31       94         94  110500000.0  565000.0     2202   
1           1   33       94         94   77000000.0  405000.0     2228   
2           2   26       92         93  118500000.0  290000.0     2143   
3           3   27       91         93   72000000.0  260000.0     1471   
4           4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  LS  ST  RS  LW  LF  CF  \
0                       5.0        4.0          4.0  90  90  90  94  95  95   
1                       5.0        4.0          5.0  94  94  94  92  93  93   
2                       5.0        5.0          5.0  87  87  87  92  92  92   
3                       4.0        3.0          1.0   0   0   0   0   0   0   
4                       4.0        5.0          4.0  85  85  85  90  90  90   

   RF  RW  LAM  CAM  RAM  LM  LCM  CM  RCM  RM  LWB  LDM  CDM  RDM  RWB  LB  \
0  95  94   95   95   95  93   86  86   86  93   66   63   63   63   66  61   
1  93  92   91   91   91  91   84  84   84  91   68   64   64   64   68  64   
2  92  92   92   92   92  91   84  84   84  91   68   63   63   63   68  63   
3   0   0    0    0    0   0    0   0    0   0    0    0    0    0    0   0   
4  90  90   91   91   91  91   90  90   90  91   80   80   80   80   80  76   

   LCB  CB  RCB  RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0   49  49   49  61      84.0       95.0             70.0          90.0   
1   56  56   56  64      84.0       94.0             89.0          81.0   
2   50  50   50  63      79.0       87.0             62.0          84.0   
3    0   0    0   0      17.0       13.0             21.0          50.0   
4   69  69   69  76      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0        48.0           22.0         94.0   
1     88.0      79.0       93.0        63.0           29.0         95.0   
2     81.0      49.0       82.0        56.0           36.0         89.0   
3     43.0      64.0       12.0        38.0           30.0         12.0   
4     90.0      75.0       91.0        76.0           61.0         87.0   

   Vision  Penalties  Composure  Marking  StandingTackle  SlidingTackle  \
0    94.0       75.0       96.0     33.0            28.0           26.0   
1    82.0       85.0       95.0     28.0            31.0           23.0   
2    87.0       81.0       94.0     27.0            24.0           33.0   
3    68.0       40.0       68.0     15.0            21.0           13.0   
4    94.0       79.0       88.0     68.0            58.0           51.0   

   GKDiving  GKHandling  GKKicking  GKPositioning  GKReflexes  Release Clause  \
0       6.0        11.0       15.0           14.0         8.0     226500000.0   
1       7.0        11.0       15.0           14.0        11.0     127100000.0   
2       9.0         9.0       15.0           15.0        11.0     228100000.0   
3      90.0        85.0

In [21]:
df['pos_RW'].value_counts()

0    17553
1      365
Name: pos_RW, dtype: int64

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17918 entries, 0 to 17917
Columns: 749 entries, Unnamed: 0 to pos_ST
dtypes: float64(40), int64(31), uint8(678)
memory usage: 21.3 MB


## Predicció

Finalment volem predir el **value** .Per això ho treim del dataset amb pop("Value") i aplicam el model.

In [23]:
val = df.pop("Value")
df

Unnamed: 0  Age  Overall  Potential      Wage  Special  \
0               0   31       94         94  565000.0     2202   
1               1   33       94         94  405000.0     2228   
2               2   26       92         93  290000.0     2143   
3               3   27       91         93  260000.0     1471   
4               4   27       91         92  355000.0     2281   
...           ...  ...      ...        ...       ...      ...   
17913       18202   19       47         65    1000.0     1307   
17914       18203   19       47         63    1000.0     1098   
17915       18204   16       47         67    1000.0     1189   
17916       18205   17       47         66    1000.0     1228   
17917       18206   16       46         66    1000.0     1321   

       International Reputation  Weak Foot  Skill Moves  LS  ST  RS  LW  LF  \
0                           5.0        4.0          4.0  90  90  90  94  95   
1                           5.0        4.0          5.0  94  94  94  92  93   
2                           5.0        5.0          5.0  87  87  87  92  92   
3                           4.0        3.0          1.0   0   0   0   0   0   
4                           4.0        5.0          4.0  85  85  85  90  90   
...                         ...        ...          ...  ..  ..  ..  ..  ..   
17913                       1.0        2.0          2.0  44  44  44  46  46   
17914                       1.0        2.0          2.0  47  47  47  41  44   
17915                       1.0        3.0          2.0  47  47  47  47  48   
17916                       1.0        3.0          2.0  49  49  49  49  48   
17917                       1.0        3.0          2.0  45  45  45  47  46   

       CF  RF  RW  LAM  CAM  RAM  LM  LCM  CM  RCM  RM  LWB  LDM  CDM  RDM  \
0      95  95  94   95   95   95  93   86  86   86  93   66   63   63   63   
1      93  93  92   91   91   91  91   84  84   84  91   68   64   64   64   
2      92  92  92   92   92   92  91   84  84   84  91   68   63   63   63   
3       0   0   0    0    0    0   0    0   0    0   0    0    0    0    0   
4      90  90  90   91   91   91  91   90  90   90  91   80   80   80   80   
...    ..  ..  ..  ...  ...  ...  ..  ...  ..  ...  ..  ...  ...  ...  ...   
17913  46  46  46   47   47   47  46   47  47   47  46   46   47   47   47   
17914  44  44  41   42   42   42  40   37  37   37  40   32   33   33   33   
17915  48  48  47   46   46   46  46   40  40   40  46   36   32   32   32   
17916  48  48  49   47   47   47  48   41  41   41  48   38   34   34   34   
17917  46  46  47   47   47   47  48   47  47   47  48   48   48   48   48   

       RWB  LB  LCB  CB  RCB  RB  Crossing  Finishing  HeadingAccuracy  \
0       66  61   49  49   49  61      84.0       95.0             70.0   
1       68  64   56  56   56  64      84.0       94.0             89.0   
2       68  63   50  50   50  63      79.0       87.0             62.0   
3        0   0    0   0    0   0      17.0       13.0             21.0   
4       80  76   69  69   69  76      93.0       82.0             55.0   
...    ...  ..  ...  ..  ...  ..       ...        ...              ...   
17913   46  47   47  47   47  47      34.0       38.0             40.0   
17914   32  31   34  34   34  31      23.0       52.0             52.0   
17915   36  35   30  30   30  35      25.0       40.0             46.0   
17916   38  37   33  33   33  37      44.0       50.0             39.0   
17917   48  48   49  49   49  48      41.0       34.0             46.0   

       ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  LongPassing  \
0              90.0     86.0       97.0   93.0        94.0         87.0   
1              81.0     87.0       88.0   81.0        76.0         77.0   
2              84.0     84.0       96.0   88.0        87.0         78.0   
3              50.0     13.0       18.0   21.0        19.0         51.0   
4              92.0     82.0       86.0   85.0        83.0         91.0   
.

Separam el data set en dos conjunts, un per fer el test i l'altre per entrenar amb el model.

In [24]:
X_train, X_test, y_train, y_test = train_test_split(df, val, test_size=0.33, random_state=42)

In [25]:
len(X_train)

12005

Entrenam un LinearRegression model.

In [26]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

Finalment obtenim una métrica $R^2$ per a la regressío, utilitz la implementació de <a href="https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html"> sickit-learn</a>.

In [27]:
preds = reg.predict(X_test)

In [28]:
preds[0]

-447513.5713709798

In [29]:
y_test[0]

110500000.0

In [30]:
r2_score(preds, y_test)

0.9672990111410212